client/relay-chain-local/src/lib.rs

```diff
@@ -0,0 +1,366 @@
+		self.full_client.import_notification_stream()
+	}
+
+	fn finality_notification_stream(&self) -> sc_client_api::FinalityNotifications<PBlock> {
+		self.full_client.finality_notification_stream()
+	}
+
+	fn storage_changes_notification_stream(
+		&self,
+		filter_keys: Option<&[sc_client_api::StorageKey]>,
+		child_filter_keys: Option<
+			&[(sc_client_api::StorageKey, Option<Vec<sc_client_api::StorageKey>>)],
+		>,
+	) -> sc_client_api::blockchain::Result<sc_client_api::StorageEventStream<PHash>> {
+		self.full_client
+			.storage_changes_notification_stream(filter_keys, child_filter_keys)
+	}
+
+	fn best_block_hash(&self) -> PHash {
+		self.backend.blockchain().info().best_hash
+	}
+
+	fn block_status(&self, block_id: BlockId) -> Result<BlockStatus, sp_blockchain::Error> {
+		self.backend.blockchain().status(block_id)
+	}
+
+	fn is_major_syncing(&self) -> bool {
+		let mut network = self.sync_oracle.lock().unwrap();
+		network.is_major_syncing()
+	}
+
+	fn overseer_handle(&self) -> Option<Handle> {
+		self.overseer_handle.clone()
+	}
+
+	fn get_storage_by_key(&self, block_id: &BlockId, key: &[u8]) -> Option<StorageValue> {
+		let state = self
+			.backend
+			.state_at(*block_id)
+			.map_err(|e| {
+				tracing::error!(
+					target: LOG_TARGET,
+					relay_parent = ?block_id,
+					error = ?e,
+					"Cannot obtain the state of the relay chain.",
+				)
+			})
+			.ok()?;
+		state
+			.storage(key)
+			.map_err(|e| {
+				tracing::error!(
+					target: LOG_TARGET,
+					error = ?e,
+					"Cannot decode the hrmp ingress channel index.",
+				)
+			})
+			.ok()?
+	}
+
+	fn prove_read(
+		&self,
+		block_id: &BlockId,
+		relevant_keys: &Vec<Vec<u8>>,
+	) -> Result<Option<StorageProof>, Box<dyn sp_state_machine::Error>> {
+		let state_backend = self
+			.backend
+			.state_at(*block_id)
+			.map_err(|e| {
+				tracing::error!(
+					target: LOG_TARGET,
+					relay_parent = ?block_id,
+					error = ?e,
+					"Cannot obtain the state of the relay chain.",
+				);
+			})
+			.ok();
+
+		match state_backend {
+			Some(state) => sp_state_machine::prove_read(state, relevant_keys)
+				.map_err(|e| {
+					tracing::error!(
+						target: LOG_TARGET,
+						relay_parent = ?block_id,
+						error = ?e,
+						"Failed to collect required relay chain state storage proof.",
+					);
+					e
+				})
+				.map(|v| Some(v)),
+			None => Ok(None),
+		}
+	}
+
+	fn check_block_in_chain(
+		&self,
+		block_id: BlockId,
+	) -> Result<BlockCheckResult, sp_blockchain::Error> {
+		let _lock = self.backend.get_import_lock();
+
+		match self.backend.blockchain().status(block_id) {
+			Ok(BlockStatus::InChain) => return Ok(BlockCheckResult::InChain),
+			Err(err) => return Err(err),
+			_ => {},
+		}
+
+		let listener = self.full_client.import_notification_stream();
+
+		// Now it is safe to drop the lock, even when the block is now imported, it should show
+		// up in our registered listener.
+		drop(_lock);
+
+		Ok(BlockCheckResult::NotFound(listener))
```

Comment:
Maybe we should instead directly return a future that will resolve when the block is imported?